src/condor_tests/ornithology/scripts/debug_plugin.py

```diff
@@ -0,0 +1,611 @@
+        "ErrorString": "Invalid plugin parameters",
+        "PluginVersion": PLUGIN_VERSION,
+        "PluginLaunched": True,
+    },
+    OPT_ERR_RESOLUTION: {
+        "ErrorCode": -2,
+        "ErrorString": "Failed to resolve server name to contact",
+        "FailedName": DEFAULT_HOSTNAME,
+        "FailureType": "Definitive",
+    },
+    OPT_ERR_CONTACT: {
+        "ErrorCode": -3,
+        "ErrorString": "Failed to contact server",
+        "FailedServer": DEFAULT_HOSTNAME,
+    },
+    OPT_ERR_AUTH: {
+        "ErrorCode": -4,
+        "ErrorString": "Failed authorization with server",
+        "FailedServer": DEFAULT_HOSTNAME,
+        "FailureType": "Authentication",
+        "ShouldRefresh": True,
+    },
+    OPT_ERR_SPECIFICATION: {
+        "ErrorCode": -5,
+        "ErrorString": "Speficied file DNE",
+        "FailedServer": DEFAULT_HOSTNAME,
+    },
+    OPT_ERR_XFER: {
+        "ErrorCode": -6,
+        "ErrorString": "Failed to transfer file",
+        "FailedServer": DEFAULT_HOSTNAME,
+        "FailureType": "ToSlow",
+    },
+}
+
+
+ACTION_TYPE_STR = {
+    ActionType.NONE: "DetailsFromDefaults",
+    ActionType.INLINE: "DetailsFromInline",
+    ActionType.FILE: "DetailsFromFile",
+    ActionType.SCRIPT: "DetailsFromScript",
+}
+
+
+# ----------------------------------------------------------------------------
+def format_error(error: Exception) -> str:
+    """Format an exception into error string"""
+    return "{0}: {1}".format(type(error).__name__, str(error))
+
+
+# ----------------------------------------------------------------------------
+def get_error_dict(error: Exception, url: str = "", parse_failure: bool = False) -> dict:
+    """Return runtime plugin failure ClassAd"""
+    error_string = format_error(error)
+    error_dict = {
+        "TransferSuccess": False,
+        "TransferError": error_string,
+        "TransferUrl": url,
+        "IsParseFailure" : parse_failure,
+    }
+
+    return error_dict
+
+
+# ----------------------------------------------------------------------------
+def parse_portion(url: str):
+    """Parse out and yield the next portion of the URL to process"""
+    option = ""
+    details = ""
+    action = ActionType.NONE
+    in_brackets = 0
+    trailing = False
+    specifier = None
+
+    # Parse URL one character at a time
+    for c in url:
+        # Always expect a '[' after an action specifier ('~','#')
+        if specifier is not None and c != "[":
+            raise DebugUrlParseError(f"Special file path delimeter token '{specifier}' not followed by '['")
+        else:
+            specifier = None
+
+        # If not in brackets and '/' then yield stored information and rest
+        if in_brackets == 0 and c == "/":
+            yield (option, action, details)
+            option = details = ""
+            action = ActionType.NONE
+            trailing = False
+        # Ignore all information post closing bracket ']'
+        elif trailing:
+            continue
```

Comment:
Wouldn't it make more sense to exit the loop here, rather than `continue` past each individual character?